### PR TITLE
One-shot decrypt directly in to destination with padding

### PR DESCRIPTION
**Pull request by vcsjones**
This is an experimental idea to see if we can make the one-shot decryption utilize the destination buffer better, but it does not come without some tradeoffs. I want to understand if others think these tradeoffs are worth it.

Today, when decrypting with the one shots, when there is any kind of padding removal, we always decrypt in to a rented buffer, then copy the resulting bytes from the rented buffer to the output buffer. This is largely because we don't know if the destination is big enough until after the decryption has already taken place.

However, if the output buffer supplied by the developer is big enough to do the transformation, we could just transform directly to the output buffer. This comes with a few trade offs:

1. The output buffer is written past what we report in `bytesWritten`, but those bytes are not material to the final answer of the decryption. They are just padding bytes. A [discussion on twitter](https://twitter.com/LeviBroderick/status/1367607494349004804) leads me to believe this is okay.
2. I'm not sure I like that in a crypto-sensitive area. Consider code like this:

    ```c#
    Aes aes = Aes.Create();
    // Set up AES
    Span<byte> destination = new byte[]; // Some overly large buffer
    int written = aes.DecryptCbc(ciphertext, iv, destination);
    // Do something with the decrypted data
    CryptographicOperations.ZeroMemory(destination.Slice(0, written));
    ```

    This decrypts in to the buffer, and the `written` amount is the actual plaintext. But there is padding in `destination` after the `written` amount that isn't getting zeroed because the developer reasonably assumed to slice destination by the amount written.

    My solution to this was to zero out the padding after the transformation. Though it does mean it's still writing beyond what was reported in `bytesWritten`, the buffer does not leak the length of the last block of data.

So, what are the benefits of this?

The biggest benefit is avoiding the copy and applying less pressure to the ArrayPool. The "tl;dr" story fairly predictable: the larger the plaintext, the better the improvement since there is no copy. For example, for a 5 MB with PKCS7 padding, it's an improvement from 611,535.8 ns down to 457,595.3 ns.

I would like to know if this is worth to continue to pursue - the benefits are not immediately clear to me except in the scenario with large data, which, I am not certain is a common use case for the one-shots. The main remaining work would be any tests, e.g. making sure padding is not left in the buffer.

<details>
<summary>Big Perf table</summary>

Benchmark code is available here: https://gist.github.com/vcsjones/209a8f263bc1ed96ecfc0c3fc4586fe3

Explanations:

1. "ExactBufferSize" are where we are forced to decrypt in to a temporary buffer.
2. "GuaranteedBufferSize" are where the destination buffer can be decrypted in to directly.

|                           Method |        Job | Toolchain | DataSize | Padding |         Mean |       Error |      StdDev | Ratio |  Gen 0 | Gen 1 | Gen 2 |
|--------------------------------- |----------- |---------- |--------- |-------- |-------------:|------------:|------------:|------:|-------:|------:|------:|
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |       16 |    None |     548.7 ns |     6.75 ns |     6.31 ns |  1.00 | 0.0200 |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |       16 |    None |     550.5 ns |     4.11 ns |     3.85 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |       16 |    None |     556.3 ns |     5.59 ns |     5.22 ns |  1.03 | 0.0200 |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |       16 |    None |     542.3 ns |     4.12 ns |     3.44 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |       16 |   PKCS7 |     587.0 ns |     4.80 ns |     4.49 ns |  0.98 | 0.0200 |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |       16 |   PKCS7 |     599.2 ns |     7.93 ns |     7.42 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |       16 |   PKCS7 |     646.6 ns |     1.56 ns |     1.46 ns |  1.07 | 0.0200 |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |       16 |   PKCS7 |     605.4 ns |     8.61 ns |     8.05 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |     1024 |    None |     629.9 ns |     6.24 ns |     5.84 ns |  0.99 | 0.0200 |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |     1024 |    None |     637.5 ns |     3.57 ns |     3.34 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |     1024 |    None |     629.0 ns |     8.74 ns |     8.17 ns |  1.00 | 0.0200 |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |     1024 |    None |     628.1 ns |     5.62 ns |     5.26 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |     1024 |   PKCS7 |     672.3 ns |     5.80 ns |     5.42 ns |  0.96 | 0.0200 |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |     1024 |   PKCS7 |     702.9 ns |     3.03 ns |     2.84 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |     1024 |   PKCS7 |     696.1 ns |     6.58 ns |     6.16 ns |  0.99 | 0.0200 |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |     1024 |   PKCS7 |     704.8 ns |     2.23 ns |     1.86 ns |  1.00 | 0.0200 |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |  1048576 |    None |  91,017.0 ns |   108.32 ns |    96.03 ns |  1.00 |      - |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |  1048576 |    None |  90,959.5 ns |   103.38 ns |    96.70 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |  1048576 |    None |  90,989.2 ns |   105.53 ns |    98.71 ns |  1.00 |      - |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |  1048576 |    None |  90,981.0 ns |    72.17 ns |    60.27 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |  1048576 |   PKCS7 |  91,056.1 ns |   137.39 ns |   128.51 ns |  0.75 |      - |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |  1048576 |   PKCS7 | 120,643.9 ns |    90.36 ns |    80.10 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |  1048576 |   PKCS7 | 120,660.5 ns |    92.72 ns |    82.19 ns |  1.00 |      - |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |  1048576 |   PKCS7 | 120,698.7 ns |   130.24 ns |   115.46 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |  5242880 |    None | 460,830.6 ns |   522.00 ns |   462.74 ns |  1.00 |      - |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |  5242880 |    None | 461,280.9 ns |   854.97 ns |   713.94 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |  5242880 |    None | 467,739.2 ns | 1,109.34 ns | 1,037.67 ns |  1.00 |      - |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |  5242880 |    None | 468,123.9 ns |   676.93 ns |   633.20 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
| Decrypt_Cbc_GuaranteedBufferSize | Job-NUITST |    branch |  5242880 |   PKCS7 | 457,595.3 ns | 1,326.97 ns | 1,241.25 ns |  0.75 |      - |     - |     - |
| Decrypt_Cbc_GuaranteedBufferSize | Job-PGYWYV |      main |  5242880 |   PKCS7 | 611,535.8 ns |   973.91 ns |   911.00 ns |  1.00 |      - |     - |     - |
|                                  |            |           |          |         |              |             |             |       |        |       |       |
|      Decrypt_Cbc_ExactBufferSize | Job-NUITST |    branch |  5242880 |   PKCS7 | 604,729.1 ns | 1,142.58 ns | 1,012.87 ns |  0.99 |      - |     - |     - |
|      Decrypt_Cbc_ExactBufferSize | Job-PGYWYV |      main |  5242880 |   PKCS7 | 611,380.8 ns | 1,667.38 ns | 1,559.66 ns |  1.00 |      - |     - |     - |
</details>

### CODE REVIEW

File: src/libraries/Common/src/Internal/Cryptography/UniversalCryptoOneShot.cs [hunk 1]
```diff
@@ -19,27 +19,49 @@ internal static class UniversalCryptoOneShot
...
+                    // The padding is invalid, but don't leave the plaintext in the buffer.
+                    CryptographicOperations.ZeroMemory(transformBuffer);
+                    throw;
+                }
+            }
 
-                // If no padding is going to be removed, we know the buffer is too small and we can bail out.
+            // If no padding is going to removed, then we already know the buffer is too small
+            // since that requires a buffer at-least the size of the ciphertext. Bail out early.
+            if (!SymmetricPadding.DepaddingRequired(paddingMode))
```
Comment (bartonjs) [inline src/libraries/Common/src/Internal/Cryptography/UniversalCryptoOneShot.cs:56]: We could also early abort if (input.Length - cipher.BlockSizeInBytes) > output.Length, since no removable symmetric padding algorithm that we support ever adds more than one cipher-block of padding.

File: src/libraries/Common/src/Internal/Cryptography/UniversalCryptoOneShot.cs [hunk 2]
```diff
@@ -19,27 +19,49 @@ internal static class UniversalCryptoOneShot
...
+                    bytesWritten = SymmetricPadding.GetPaddingLength(transformBuffer, paddingMode, cipher.BlockSizeInBytes);
+
+                    // Zero out the padding so that the buffer does not contain the padding data "after" the bytesWritten.
+                    CryptographicOperations.ZeroMemory(transformBuffer.Slice(bytesWritten));
                     return true;
                 }
+                catch (CryptographicException)
+                {
+                    // The padding is invalid, but don't leave the plaintext in the buffer.
+                    CryptographicOperations.ZeroMemory(transformBuffer);
```
Comment (bartonjs) [inline src/libraries/Common/src/Internal/Cryptography/UniversalCryptoOneShot.cs:49]: Once we're willing to let "unverified" data be written into the destination buffer, we could do all of this differently.

* If output.Length >= input.Length, do this.
* else if no removable padding, fail
* else if output.Length < input.Length - BlockSizeInBytes, fail
* try
  * Decrypt input.Length - BlockSizeInBytes into output.
  * stackalloc 16 (Debug.Assert BlockSizeInBytes <= stackalloc.Length)
  * Decrypt the last block into the stackalloc
  * Apply depadding, copy or fail as appropriate
* catch
  * clear output
  * throw-continue

And that eliminates the rented array entirely, at the cost of a second P/Invoke for the "it might or might not fit" case.

Historically, I'm not a fan of using caller provided buffers as scratch space; but it happens in enough places that I may as well embrace it.

Comment (vcsjones) [reply src/libraries/Common/src/Internal/Cryptography/UniversalCryptoOneShot.cs:49]: Hm. This gets a little ugly with CFB but, I think I can make it work.

**Comment by vcsjones** [comment]
@bartonjs I pushed up some changes based on your feedback. New benchmarks below.

tl;dr: the large data sizes increased as one would expect due to the less copying. Small data regressed, but, I shortcut those to do a single call in a stack buffer which makes small data (<=128) "happy" again.


<details>

|                           Method |        Job | Toolchain | DataSize | Padding |         Mean |       Error |      StdDev | Ratio | RatioSD |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------- |----------- |---------- |--------- |-------- |-------------:|------------:|------------:|------:|--------:|-------:|------:|------:|----------:|
| Decrypt_Cbc_GuaranteedBufferSize | Job-SSHGXV |    branch |       16 |   PKCS7 |     580.2 ns |     6.74 ns |     6.30 ns |  0.96 |    0.01 | 0.0200 |     - |     - |     128 B |
| Decrypt_Cbc_GuaranteedBufferSize | Job-ZZAIDD |      main |       16 |   PKCS7 |     602.1 ns |     3.56 ns |     3.16 ns |  1.00 |    0.00 | 0.0200 |     - |     - |     128 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
|      Decrypt_Cbc_ExactBufferSize | Job-SSHGXV |    branch |       16 |   PKCS7 |     564.5 ns |     5.03 ns |     4.70 ns |  0.94 |    0.01 | 0.0200 |     - |     - |     128 B |
|      Decrypt_Cbc_ExactBufferSize | Job-ZZAIDD |      main |       16 |   PKCS7 |     598.6 ns |     4.96 ns |     4.64 ns |  1.00 |    0.00 | 0.0200 |     - |     - |     128 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
| Decrypt_Cbc_GuaranteedBufferSize | Job-SSHGXV |    branch |      128 |   PKCS7 |     593.6 ns |     4.50 ns |     4.21 ns |  1.00 |    0.01 | 0.0200 |     - |     - |     128 B |
| Decrypt_Cbc_GuaranteedBufferSize | Job-ZZAIDD |      main |      128 |   PKCS7 |     590.7 ns |     6.75 ns |     6.32 ns |  1.00 |    0.00 | 0.0200 |     - |     - |     128 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
|      Decrypt_Cbc_ExactBufferSize | Job-SSHGXV |    branch |      128 |   PKCS7 |     602.4 ns |     4.33 ns |     3.84 ns |  1.00 |    0.01 | 0.0200 |     - |     - |     128 B |
|      Decrypt_Cbc_ExactBufferSize | Job-ZZAIDD |      main |      128 |   PKCS7 |     602.2 ns |     3.65 ns |     3.23 ns |  1.00 |    0.00 | 0.0200 |     - |     - |     128 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
| Decrypt_Cbc_GuaranteedBufferSize | Job-SSHGXV |    branch |      129 |   PKCS7 |     594.3 ns |     7.26 ns |     6.79 ns |  1.00 |    0.02 | 0.0200 |     - |     - |     128 B |
| Decrypt_Cbc_GuaranteedBufferSize | Job-ZZAIDD |      main |      129 |   PKCS7 |     593.8 ns |     9.23 ns |     8.63 ns |  1.00 |    0.00 | 0.0200 |     - |     - |     128 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
|      Decrypt_Cbc_ExactBufferSize | Job-SSHGXV |    branch |      129 |   PKCS7 |     609.0 ns |     3.75 ns |     3.50 ns |  1.00 |    0.01 | 0.0200 |     - |     - |     128 B |
|      Decrypt_Cbc_ExactBufferSize | Job-ZZAIDD |      main |      129 |   PKCS7 |     608.1 ns |     2.92 ns |     2.59 ns |  1.00 |    0.00 | 0.0200 |     - |     - |     128 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
| Decrypt_Cbc_GuaranteedBufferSize | Job-SSHGXV |    branch |  5242880 |   PKCS7 | 460,490.4 ns |   952.68 ns |   795.53 ns |  0.74 |    0.01 |      - |     - |     - |     128 B |
| Decrypt_Cbc_GuaranteedBufferSize | Job-ZZAIDD |      main |  5242880 |   PKCS7 | 617,425.1 ns | 7,254.54 ns | 6,785.90 ns |  1.00 |    0.00 |      - |     - |     - |     129 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
|      Decrypt_Cbc_ExactBufferSize | Job-SSHGXV |    branch |  5242880 |   PKCS7 | 476,455.1 ns | 2,884.06 ns | 2,556.64 ns |  0.78 |    0.00 |      - |     - |     - |     128 B |
|      Decrypt_Cbc_ExactBufferSize | Job-ZZAIDD |      main |  5242880 |   PKCS7 | 606,969.7 ns | 1,131.37 ns | 1,002.93 ns |  1.00 |    0.00 |      - |     - |     - |     129 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
| Decrypt_Cbc_GuaranteedBufferSize | Job-SSHGXV |    branch |  5242881 |   PKCS7 | 462,148.7 ns | 2,065.94 ns | 1,831.40 ns |  0.76 |    0.00 |      - |     - |     - |     128 B |
| Decrypt_Cbc_GuaranteedBufferSize | Job-ZZAIDD |      main |  5242881 |   PKCS7 | 605,905.5 ns | 1,016.27 ns |   848.63 ns |  1.00 |    0.00 |      - |     - |     - |     129 B |
|                                  |            |           |          |         |              |             |             |       |         |        |       |       |           |
|      Decrypt_Cbc_ExactBufferSize | Job-SSHGXV |    branch |  5242881 |   PKCS7 | 461,642.1 ns |   539.61 ns |   450.60 ns |  0.75 |    0.00 |      - |     - |     - |     128 B |
|      Decrypt_Cbc_ExactBufferSize | Job-ZZAIDD |      main |  5242881 |   PKCS7 | 614,144.4 ns |   732.79 ns |   611.91 ns |  1.00 |    0.00 |      - |     - |     - |     129 B |
</details>

**Comment by vcsjones** [comment]
... but it passes for me locally..

**Comment by vcsjones** [comment]
Okay. I think prior to this, we (I) need to lite-ify `AesCng` and `TripleDESCng`. Their one shot implementation is in `UniversalCryptoDecryptor` because they needed to be able to handle ncrypt keys, so it has a duplicate implementation of the one shot.

Since it sounds like we're open to "decrypt in to the user buffer" I will re-open a PR for that, but I would take to take care of the one-shot implementation on the public CNG types, first.
